File: backend/app/services/validaciones.py
"""
Logica de validacion de horarios.

Todas las funciones de este modulo son puras: reciben los datos como
argumentos y devuelven el resultado sin modificar nada ni tocar la base
de datos. Eso permite probarlas de forma aislada con pytest y es el
enfoque que estamos trabajando en la materia (funciones puras,
inmutabilidad y composicion).
"""


def a_minutos(hora: str) -> int:
    """Convierte 'HH:MM' a minutos desde medianoche."""
    h, m = hora.split(":")
    return int(h) * 60 + int(m)


def hay_traslape(inicio1: str, fin1: str, inicio2: str, fin2: str) -> bool:
    """Dos intervalos se traslapan si uno empieza antes de que termine el otro."""
    return a_minutos(inicio1) < a_minutos(fin2) and a_minutos(fin1) > a_minutos(inicio2)


def duracion_horas(inicio: str, fin: str) -> float:
    return (a_minutos(fin) - a_minutos(inicio)) / 60


def _bloque_texto(propuesta: dict) -> str:
    return f"{propuesta['asignatura_id']}/{propuesta['paralelo_id']} {propuesta['dia_semana']} {propuesta['hora_inicio']}-{propuesta['hora_fin']}"


def _conflicto(codigo: str, detalle: str, propuesta: dict) -> dict:
    return {"codigo": codigo, "detalle": detalle, "bloque": _bloque_texto(propuesta)}


# ------------------------------------------------------------------
# Validaciones individuales: cada una devuelve una lista de conflictos
# (vacia si no hay problema)
# ------------------------------------------------------------------

def validar_distributivo(propuesta: dict, distributivo: list[dict]) -> list[dict]:
    existe = any(
        d["docente_id"] == propuesta["docente_id"]
        and d["asignatura_id"] == propuesta["asignatura_id"]
        and d["paralelo_id"] == propuesta["paralelo_id"]
        for d in distributivo
    )
    if existe:
        return []
    return [_conflicto(
        "DISTRIBUTIVO_INVALIDO",
        "La combinacion docente/asignatura/paralelo no existe en el distributivo",
        propuesta,
    )]


def validar_disponibilidad_docente(propuesta: dict, disponibilidades: list[dict]) -> list[dict]:
    franjas = [
        d for d in disponibilidades
        if d["docente_id"] == propuesta["docente_id"]
        and d["dia_semana"] == propuesta["dia_semana"]
        and d["disponible"]
    ]
    cubierto = any(
        a_minutos(f["hora_inicio"]) <= a_minutos(propuesta["hora_inicio"])
        and a_minutos(f["hora_fin"]) >= a_minutos(propuesta["hora_fin"])
        for f in franjas
    )
    if cubierto:
        return []
    return [_conflicto(
        "FUERA_DISPONIBILIDAD",
        f"El docente {propuesta['docente_id']} no tiene disponibilidad en esa franja",
        propuesta,
    )]


def validar_docente_ocupado(propuesta: dict, horarios: list[dict]) -> list[dict]:
    ocupado = [
        h for h in horarios
        if h["docente_id"] == propuesta["docente_id"]
        and h["dia_semana"] == propuesta["dia_semana"]
        and hay_traslape(propuesta["hora_inicio"], propuesta["hora_fin"], h["hora_inicio"], h["hora_fin"])
    ]
    return [
        _conflicto(
            "DOCENTE_OCUPADO",
            f"El docente {propuesta['docente_id']} ya dicta {h['asignatura_id']} en esa franja",
            propuesta,
        )
        for h in ocupado
    ]


def validar_espacio_ocupado(propuesta: dict, horarios: list[dict]) -> list[dict]:
    ocupado = [
        h for h in horarios
        if h["espacio_id"] == propuesta["espacio_id"]
        and h["dia_semana"] == propuesta["dia_semana"]
        and hay_traslape(propuesta["hora_inicio"], propuesta["hora_fin"], h["hora_inicio"], h["hora_fin"])
    ]
    return [
        _conflicto(
            "ESPACIO_OCUPADO",
            f"El espacio {propuesta['espacio_id']} ya esta ocupado por {h['asignatura_id']} en esa franja",
            propuesta,
        )
        for h in ocupado
    ]


def validar_espacio_compatible(propuesta: dict, asignaturas: list[dict], espacios: list[dict]) -> list[dict]:
    asignatura = next((a for a in asignaturas if a["asignatura_id"] == propuesta["asignatura_id"]), None)
    espacio = next((e for e in espacios if e["espacio_id"] == propuesta["espacio_id"]), None)

    if asignatura is None or espacio is None:
        return [_conflicto("DATO_INEXISTENTE", "La asignatura o el espacio no estan registrados", propuesta)]

    requerido = asignatura.get("tipo_espacio_requerido")
    if not requerido or requerido == espacio["tipo_espacio"]:
        return []
    return [_conflicto(
        "ESPACIO_INCOMPATIBLE",
        f"La asignatura requiere {requerido} y el espacio es {espacio['tipo_espacio']}",
        propuesta,
    )]


def validar_carga_horaria(propuesta: dict, horarios: list[dict], docentes: list[dict]) -> list[dict]:
    docente = next((d for d in docentes if d["docente_id"] == propuesta["docente_id"]), None)
    if docente is None:
        return [_conflicto("DATO_INEXISTENTE", "El docente no esta registrado", propuesta)]

    horas_actuales = sum(
        duracion_horas(h["hora_inicio"], h["hora_fin"])
        for h in horarios
        if h["docente_id"] == propuesta["docente_id"]
    )
    horas_nuevas = duracion_horas(propuesta["hora_inicio"], propuesta["hora_fin"])
    maximo = docente["horas_max_semanales"]

    if horas_actuales + horas_nuevas <= maximo:
        return []
    return [_conflicto(
        "EXCESO_CARGA",
        f"Con este bloque el docente llegaria a {horas_actuales + horas_nuevas:.1f} horas y su maximo es {maximo}",
        propuesta,
    )]


def validar_capacidad(propuesta: dict, paralelos: list[dict], espacios: list[dict]) -> list[dict]:
    paralelo = next((p for p in paralelos if p["paralelo_id"] == propuesta["paralelo_id"]), None)
    espacio = next((e for e in espacios if e["espacio_id"] == propuesta["espacio_id"]), None)

    if paralelo is None or espacio is None:
        return [_conflicto("DATO_INEXISTENTE", "El paralelo o el espacio no estan registrados", propuesta)]

    if paralelo["numero_estudiantes"] <= espacio["capacidad"]:
        return []
    return [_conflicto(
        "CAPACIDAD_INSUFICIENTE",
        f"El paralelo tiene {paralelo['numero_estudiantes']} estudiantes y el espacio soporta {espacio['capacidad']}",
        propuesta,
    )]


# ------------------------------------------------------------------
# Composicion: se aplican todas las validaciones y se juntan los
# resultados. La lista de funciones actua como pipeline.
# ------------------------------------------------------------------

def validar_propuesta(propuesta: dict, datos: dict) -> dict:
    """
    datos es un diccionario con los catalogos:
    docentes, espacios, asignaturas, paralelos, distributivo,
    disponibilidades y horarios (bloques ya confirmados).
    """
    validaciones = [
        lambda: validar_distributivo(propuesta, datos["distributivo"]),
        lambda: validar_disponibilidad_docente(propuesta, datos["disponibilidades"]),
        lambda: validar_docente_ocupado(propuesta, datos["horarios"]),
        lambda: validar_espacio_ocupado(propuesta, datos["horarios"]),
        lambda: validar_espacio_compatible(propuesta, datos["asignaturas"], datos["espacios"]),
        lambda: validar_carga_horaria(propuesta, datos["horarios"], datos["docentes"]),
        lambda: validar_capacidad(propuesta, datos["paralelos"], datos["espacios"]),
    ]

    conflictos = [c for validar in validaciones for c in validar()]

    return {
        "estado": "INVALIDO" if conflictos else "VALIDO",
        "total_conflictos": len(conflictos),
        "conflictos": conflictos,
    }

File: frontend/src/app/servicios/validaciones.ts
/**
 * Espejo en TypeScript de backend/app/services/validaciones.py
 *
 * Se usa unicamente en el modo demostracion (publicacion en GitHub Pages),
 * donde no hay backend disponible y la validacion se resuelve en el navegador.
 * En la ejecucion normal del sistema estas reglas se evaluan en el backend y,
 * al confirmar un bloque, tambien en el procedimiento almacenado de SQL Server.
 *
 * Al igual que en el backend, todas las funciones son puras: reciben los datos
 * como argumentos, no modifican nada y devuelven la lista de conflictos.
 */

export interface Conflicto {
  codigo: string;
  detalle: string;
  bloque: string;
}

export interface ResultadoValidacion {
  estado: string;
  total_conflictos: number;
  conflictos: Conflicto[];
}

export function aMinutos(hora: string): number {
  const [h, m] = hora.split(':').map(Number);
  return h * 60 + m;
}

export function hayTraslape(inicio1: string, fin1: string, inicio2: string, fin2: string): boolean {
  return aMinutos(inicio1) < aMinutos(fin2) && aMinutos(fin1) > aMinutos(inicio2);
}

export function duracionHoras(inicio: string, fin: string): number {
  return (aMinutos(fin) - aMinutos(inicio)) / 60;
}

function textoBloque(p: any): string {
  return `${p.asignatura_id}/${p.paralelo_id} ${p.dia_semana} ${p.hora_inicio}-${p.hora_fin}`;
}

function conflicto(codigo: string, detalle: string, p: any): Conflicto {
  return { codigo, detalle, bloque: textoBloque(p) };
}

// ---------------- validaciones individuales ----------------

export function validarDistributivo(p: any, distributivo: any[]): Conflicto[] {
  const existe = distributivo.some(d =>
    d.docente_id === p.docente_id &&
    d.asignatura_id === p.asignatura_id &&
    d.paralelo_id === p.paralelo_id
  );
  return existe ? [] : [conflicto(
    'DISTRIBUTIVO_INVALIDO',
    'La combinacion docente/asignatura/paralelo no existe en el distributivo',
    p
  )];
}

export function validarDisponibilidadDocente(p: any, disponibilidades: any[]): Conflicto[] {
  const cubierto = disponibilidades.some(d =>
    d.docente_id === p.docente_id &&
    d.dia_semana === p.dia_semana &&
    d.disponible &&
    aMinutos(d.hora_inicio) <= aMinutos(p.hora_inicio) &&
    aMinutos(d.hora_fin) >= aMinutos(p.hora_fin)
  );
  return cubierto ? [] : [conflicto(
    'FUERA_DISPONIBILIDAD',
    `El docente ${p.docente_id} no tiene disponibilidad en esa franja`,
    p
  )];
}

export function validarDocenteOcupado(p: any, horarios: any[]): Conflicto[] {
  return horarios
    .filter(h =>
      h.docente_id === p.docente_id &&
      h.dia_semana === p.dia_semana &&
      hayTraslape(p.hora_inicio, p.hora_fin, h.hora_inicio, h.hora_fin))
    .map(h => conflicto(
      'DOCENTE_OCUPADO',
      `El docente ${p.docente_id} ya dicta ${h.asignatura_id} en esa franja`,
      p
    ));
}

export function validarEspacioOcupado(p: any, horarios: any[]): Conflicto[] {
  return horarios
    .filter(h =>
      h.espacio_id === p.espacio_id &&
      h.dia_semana === p.dia_semana &&
      hayTraslape(p.hora_inicio, p.hora_fin, h.hora_inicio, h.hora_fin))
    .map(h => conflicto(
      'ESPACIO_OCUPADO',
      `El espacio ${p.espacio_id} ya esta ocupado por ${h.asignatura_id} en esa franja`,
      p
    ));
}

export function validarEspacioCompatible(p: any, asignaturas: any[], espacios: any[]): Conflicto[] {
  const asignatura = asignaturas.find(a => a.asignatura_id === p.asignatura_id);
  const espacio = espacios.find(e => e.espacio_id === p.espacio_id);

  if (!asignatura || !espacio) {
    return [conflicto('DATO_INEXISTENTE', 'La asignatura o el espacio no estan registrados', p)];
  }

  const requerido = asignatura.tipo_espacio_requerido;
  if (!requerido || requerido === espacio.tipo_espacio) return [];

  return [conflicto(
    'ESPACIO_INCOMPATIBLE',
    `La asignatura requiere ${requerido} y el espacio es ${espacio.tipo_espacio}`,
    p
  )];
}

export function validarCargaHoraria(p: any, horarios: any[], docentes: any[]): Conflicto[] {
  const docente = docentes.find(d => d.docente_id === p.docente_id);
  if (!docente) {
    return [conflicto('DATO_INEXISTENTE', 'El docente no esta registrado', p)];
  }

  const horasActuales = horarios
    .filter(h => h.docente_id === p.docente_id)
    .reduce((suma, h) => suma + duracionHoras(h.hora_inicio, h.hora_fin), 0);

  const total = horasActuales + duracionHoras(p.hora_inicio, p.hora_fin);
  const maximo = docente.horas_max_semanales;

  if (total <= maximo) return [];

  return [conflicto(
    'EXCESO_CARGA',
    `Con este bloque el docente llegaria a ${total.toFixed(1)} horas y su maximo es ${maximo}`,
    p
  )];
}

export function validarCapacidad(p: any, paralelos: any[], espacios: any[]): Conflicto[] {
  const paralelo = paralelos.find(x => x.paralelo_id === p.paralelo_id);
  const espacio = espacios.find(e => e.espacio_id === p.espacio_id);

  if (!paralelo || !espacio) {
    return [conflicto('DATO_INEXISTENTE', 'El paralelo o el espacio no estan registrados', p)];
  }

  if (paralelo.numero_estudiantes <= espacio.capacidad) return [];

  return [conflicto(
    'CAPACIDAD_INSUFICIENTE',
    `El paralelo tiene ${paralelo.numero_estudiantes} estudiantes y el espacio soporta ${espacio.capacidad}`,
    p
  )];
}

// ---------------- composicion ----------------

export function validarPropuesta(p: any, datos: any): ResultadoValidacion {
  const validaciones: Array<() => Conflicto[]> = [
    () => validarDistributivo(p, datos.distributivo),
    () => validarDisponibilidadDocente(p, datos.disponibilidades),
    () => validarDocenteOcupado(p, datos.horarios),
    () => validarEspacioOcupado(p, datos.horarios),
    () => validarEspacioCompatible(p, datos.asignaturas, datos.espacios),
    () => validarCargaHoraria(p, datos.horarios, datos.docentes),
    () => validarCapacidad(p, datos.paralelos, datos.espacios)
  ];

  const conflictos = validaciones.flatMap(validar => validar());

  return {
    estado: conflictos.length > 0 ? 'INVALIDO' : 'VALIDO',
    total_conflictos: conflictos.length,
    conflictos
  };
}
